### PR TITLE
Update R2 Error Handling for Invalid Characters

### PR DESCRIPTION
**Pull request by dbenCF**
Fixes # [insert GH or internal issue number(s)].

**What this PR solves / how to test:**
Adds client side error handling for creating an R2 bucket with invalid characters
**Author has addressed the following:**

- Tests
  - [x] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [x] Included
  - [ ] Not necessary because:
- Associated docs
  - [x] Issue(s)/PR(s):https://github.com/cloudflare/cloudflare-docs/pull/14355
  - [ ] Not necessary because:

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

File: packages/wrangler/src/r2/helpers.ts [hunk 1]
```diff
@@ -299,3 +299,15 @@ export async function putR2Sippy(
...
+
+
+/**
+ * R2 bucket names must only contain alphanumeric and - characters.
+ */
+export function isValidR2BucketName(
+	name: string | undefined
+): name is string {
+	return (
+		typeof name === "string" && /^[a-zA-Z][a-zA-Z0-9-]*$/.test(name)
```
Comment (penalosa) [inline packages/wrangler/src/r2/helpers.ts:311]: Is this mentioned in the R2 documentation? If so, it'd be good to link in a comment

Comment (dbenCF) [reply packages/wrangler/src/r2/helpers.ts:311]: Not currently, however I can request this to be added to the documentation too.

File: .changeset/brown-ties-deliver.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": major
```
[the remark this answers is not in the document]
Comment (dbenCF) [reply .changeset/brown-ties-deliver.md]: Thanks for flagging, changed to patch.

File: packages/wrangler/src/r2/helpers.ts [hunk 3]
```diff
@@ -299,3 +299,14 @@ export async function putR2Sippy(
...
 }
+
+/**
+ * R2 bucket names must only contain alphanumeric and - characters.
+ */
+export function isValidR2BucketName(
+	name: string | undefined
+): name is string {
+	return (
+		typeof name === "string" && /^[a-zA-Z][a-zA-Z0-9-]*$/.test(name)
```
Comment (penalosa) [inline packages/wrangler/src/r2/helpers.ts:310]:
```suggestion
		typeof name === "string" && /^[a-zA-Z][a-zA-Z0-9-]+$/.test(name)
```

**Review by penalosa — COMMENTED** [review]
This looks good! The only thing remaining is an update to the docs—it looks like you checked off the checkbox for this including docs, but didn't link the issue/PR. Would you be able to do that, and then we can get this merged and released?

File: packages/wrangler/src/r2/helpers.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/wrangler/src/r2/helpers.ts:560]: Can a bucket not have a single character for its name?

**Review by petebacondarwin — APPROVED** [review]

File: packages/wrangler/src/r2/helpers.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/wrangler/src/r2/helpers.ts]: https://github.com/cloudflare/cloudflare-docs/pull/14355

**Comment by petebacondarwin** [comment]
I ran the E2E tests locally.

**Pull request merged**